### PR TITLE
KFSPTS-5732 Add changes so that foreign vendors with payment reason i…

### PR DESCRIPTION
**Pull request by db483**
…n new sys param PAYMENT_REASONS_THAT_DO_NOT_REQUIRE_TAX_REVIEW_FOR_FOREIGN_VENDOR don't route for Tax Review and Income Class on Nonresident Alien Tax tab is set to N

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package edu.cornell.kfs.fp.document.service;
+
+public interface CuDisbursementVoucherTaxService {
+	
+	public boolean isForeignVendorWithTaxReviewRequired(String payeeTypeCode,
+			String paymentReasonCode, Integer vendorHeaderId);
+	
+	public boolean isForeignVendoAndNOTaxReviewRequired(String payeeTypeCode,
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java]: Vendor is misspelled in method name; please correct that. Also, it might be clearer to rename the methods to something like "isForeignVendorAndTaxReviewIsRequired" for the first and "isForeignVendorAndTaxReviewIsNotRequired" for the second.

File: src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java [hunk 2]
```diff
@@ -80,4 +86,48 @@ protected boolean validateNRATaxInformation(DisbursementVoucherDocument document
         }
         return true;
     }
+    
+	public boolean isForeignVendorWithTaxReviewRequired(String payeeTypeCode,
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java]: I would recommend adding the "@Override" annotation to any overridden or implemented methods, to help minimize errors when future code changes are made.

File: src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java (hunk 1 above)
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java]: thank you, I will add the changes

File: src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java (hunk 2 above)
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java]: Will do

File: src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java [hunk 3]
```diff
@@ -80,4 +86,48 @@ protected boolean validateNRATaxInformation(DisbursementVoucherDocument document
...
+
+		return isForeignVendor(payeeTypeCode, vendorHeaderId) && paymentReasonDoesNotRequireTaxReviewForForeignVendor(paymentReasonCode);
+	}
+	
+	protected boolean isForeignVendor(String payeeTypeCode, Integer vendorHeaderId){
+		if (payeeTypeCode.equals(KFSConstants.PaymentPayeeTypes.VENDOR) && vendorService.isVendorForeign(vendorHeaderId)){ 
+			return true;
+		}
+		
+		return false;
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java]: In this method and the one below it, I would suggest returning the result of the condition so that you don't need the extra "return true/false" statements.

File: src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java [hunk 4]
```diff
@@ -80,4 +86,48 @@ protected boolean validateNRATaxInformation(DisbursementVoucherDocument document
...
+	}
+	
+	protected boolean paymentReasonDoesNotRequireTaxReviewForForeignVendor(String paymentReasonCode){
+		if (parameterEvaluatorService.getParameterEvaluator(DisbursementVoucherDocument.class, CuDisbursementVoucherConstants.PAYMENT_REASONS_THAT_DO_NOT_REQUIRE_TAX_REVIEW_FOR_FOREIGN_VENDOR, paymentReasonCode).evaluationSucceeds()) {
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java]: Parameter evaluators are capable of treating a parameter's values as a whitelist or blacklist depending on the parameter's Allow/Deny flag. There are other places in KFS that rely on this flag for treating the values as a blacklist. Your code should work as-is, but if you want to make things consistent with similar setup in KFS, you could change this around so that evaluationSucceeds() means tax review IS required, and then set the parameter's flag to Deny so that its elements will be treated as a blacklist. (This may also require a parameter name change for consistency.) Let me know if this makes sense in the event that you do intend to go forward with the change.

Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java]: This is a good point and I have considered this option. However I find it that this more clearly implements the requirements and would like to keep it as is unless there is a major concern about it.

File: src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+	
+	@Override
+	protected void setUp() throws Exception {
+		super.setUp();
+		disbursementVoucherTaxService = new CuDisbursementVoucherTaxServiceImpl();
+		disbursementVoucherTaxService.setVendorService(new MockVendorService());
+		disbursementVoucherTaxService.setParameterEvaluatorService(new MockParameterEvaluatorService());	
+	}
+	
+	public void testIsForeignVendorWithTaxReviewRequired_true()  {
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java]: You may need to rename these various test methods if you end up changing the CuDisbursementVoucherTaxService method names.

File: src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+import org.kuali.kfs.vnd.businessobject.VendorHeader;
+import org.kuali.kfs.vnd.businessobject.VendorRoutingComparable;
+import org.kuali.kfs.vnd.document.service.VendorService;
+import org.kuali.rice.core.api.parameter.ParameterEvaluator;
+import org.kuali.rice.core.api.parameter.ParameterEvaluatorService;
+import org.kuali.rice.core.api.util.type.KualiDecimal;
+import org.kuali.rice.krad.bo.Note;
+import org.kuali.rice.krad.document.Document;
+
+public class CuDisbursementVoucherTaxServiceImplTest extends TestCase {
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java]: I would suggest adding a few test cases where the DV payee is not a vendor (such as an employee payee).

File: src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+			return false;
+		}
+
+		@Override
+		public boolean isVendorForeign(Integer vendorHeaderGeneratedIdentifier) {
+			if(FOREIGN_VENDOR_HEADER_ID.equals(vendorHeaderGeneratedIdentifier)){
+				return true;
+			}
+			else
+				return false;
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java]: Another place where the return statement can be consolidated.

File: src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java (hunk 4 above)
Comment (cah292) [inline src/main/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImpl.java]: Ok, that makes sense. I would recommend tweaking this method, though, so that it returns the result of the condition. That way, you don't need the extra "return true/false" statements.

**Comment by cah292** [comment]
Changes look good in general, but I did have some suggestions for improvements. Will wait for feedback before reviewing again.

File: src/main/java/edu/cornell/kfs/fp/document/CuDisbursementVoucherDocument.java [hunk 8]
```diff
@@ -923,6 +924,48 @@ protected boolean isCampusReviewRequired() {
 
         return false;
     }
+    
+    @Override
+    protected boolean isTaxReviewRequired() {
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/fp/document/CuDisbursementVoucherDocument.java]: Since the bulk of this method is the same as the one it overrides, it would be nice to eliminate the duplication and call super.isTaxReviewRequired(), however that looked like it might be difficult based on my quick review. Given that, I'm not asking for this to be changed, but I wanted to bring it up so others can think along the same lines (avoid duplicating code from overridden methods if possible) and just in case you can see a better way to do that here than I can.

File: src/main/java/edu/cornell/kfs/fp/document/CuDisbursementVoucherDocument.java [hunk 9]
```diff
@@ -923,6 +924,48 @@ protected boolean isCampusReviewRequired() {
...
+        if (isPayeePurchaseOrderVendorHasWithholding()) {
+            return true;
+        }
+
+        String payeeTypeCode = this.getDvPayeeDetail().getDisbursementVoucherPayeeTypeCode();
+        if (payeeTypeCode.equals(KFSConstants.PaymentPayeeTypes.EMPLOYEE)) {
+            return false;
+        } else if (payeeTypeCode.equals(KFSConstants.PaymentPayeeTypes.VENDOR)) {
+            if(vendorService.isVendorInstitutionEmployee(this.getDvPayeeDetail().getDisbVchrVendorHeaderIdNumberAsInteger())){
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/fp/document/CuDisbursementVoucherDocument.java]: I know this was basically copied from the overridden method, but it makes me nervous that it's referencing the class variable rather than the getter. Since this service can't be injected, we're depending on the getter being called before this or we'll get an NPE. I would feel more comfortable with it calling the getter.

File: src/main/java/edu/cornell/kfs/fp/document/CuDisbursementVoucherDocument.java [hunk 10]
```diff
@@ -923,6 +924,48 @@ protected boolean isCampusReviewRequired() {
...
+                return true;
+            }
+        }
+        
+        String paymentReasonCode = this.getDvPayeeDetail().getDisbVchrPaymentReasonCode();
+        Integer vendorHeaderId = getDvPayeeDetail().getDisbVchrVendorHeaderIdNumberAsInteger();
+        if (SpringContext.getBean(CuDisbursementVoucherTaxServiceImpl.class).isForeignVendorWithTaxReviewRequired(payeeTypeCode, paymentReasonCode, vendorHeaderId)) {
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/fp/document/CuDisbursementVoucherDocument.java]: Would be nice to extract this to a getter for better testability.

File: src/main/java/edu/cornell/kfs/fp/document/CuDisbursementVoucherDocument.java [hunk 11]
```diff
@@ -923,6 +924,48 @@ protected boolean isCampusReviewRequired() {
...
+       
+        if (this.getDvPymentReasonService().isDecedentCompensationPaymentReason(paymentReasonCode)) {
+            return true;
+        }
+
+        if (this.getDvPymentReasonService().isMovingPaymentReason(paymentReasonCode) && taxedCampusForMovingReimbursements()) {
+            return true;
+        }
+
+        if (/*REFACTORME*/SpringContext.getBean(ParameterEvaluatorService.class).getParameterEvaluator(this.getClass(), DisbursementVoucherDocument.PAYMENT_REASONS_REQUIRING_TAX_REVIEW_PARAMETER_NAME, paymentReasonCode).evaluationSucceeds()) {
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/fp/document/CuDisbursementVoucherDocument.java]: Same as above, would be nice to extract this to a getter for better testability.

File: src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package edu.cornell.kfs.fp.document.service;
+
+public interface CuDisbursementVoucherTaxService {
+	
+	public boolean isForeignVendorWithTaxReviewRequired(String payeeTypeCode,
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java]: since this is an interface, the public access modifier is redundant and can be removed

File: src/main/java/edu/cornell/kfs/fp/document/validation/impl/CuDisbursementVoucherDocumentPreRules.java [hunk 13]
```diff
@@ -67,11 +70,26 @@ public boolean doPrompts(Document document) {
...
-    /**
+	private void setIncomeClassNonReportableForForeignVendorWithNoTaxReviewRequired(Document document) {
+		DisbursementVoucherDocument dvDoc = (DisbursementVoucherDocument) document;
+		DisbursementVoucherPayeeDetail dvPayeeDetail = dvDoc.getDvPayeeDetail();
+		
+		String payeeTypeCode = dvPayeeDetail.getDisbursementVoucherPayeeTypeCode();
+		String paymentReasonCode = dvPayeeDetail.getDisbVchrPaymentReasonCode();
+		Integer vendorHeaderId = dvPayeeDetail.getDisbVchrVendorHeaderIdNumberAsInteger();
+
+		if (SpringContext.getBean(CuDisbursementVoucherTaxServiceImpl.class).isForeignVendoAndNOTaxReviewRequired(payeeTypeCode,paymentReasonCode, vendorHeaderId)) {
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/fp/document/validation/impl/CuDisbursementVoucherDocumentPreRules.java]: Another place where SpringContext.getBean() could be extracted into a getter for better testability.

File: src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java (hunk 6 above)
Comment (bhutchinson) [inline src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java]: FWIW, using JUnit 4, you don't need to extend TestCase, and you can use annotations - @Before for setup, @After for teardown and @Test for test methods. You can look at edu.cornell.kfs.pdp.batch.service.impl.PayeeACHAccountExtractServiceImplTest for an example.

File: src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java (hunk 1 above)
Comment (se12) [inline src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java]: Does the additional 'Is' in the proposed method name necessary?  It seems grammatically incorrect, does not flow right in my mind.

Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java]: Sandy, would it make more sense to rename them to: 
"isForeignVendorAndTaxReviewRequired" and "isForeignVendorAndTaxReviewNotRequired"? Thank you.

File: src/main/java/edu/cornell/kfs/fp/document/CuDisbursementVoucherDocument.java (hunk 8 above)
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/document/CuDisbursementVoucherDocument.java]: Thank you for pointing this out Bryan. Unfortunately I could not find a way to call super and eliminate duplication. I know this will make it difficult when we upgrade. If this method changes in foundation code it can be tricky to determine what was Cornell customization and what not, and the developer doing the upgrade would need to understand what changed in foundation code between different versions and what part was the intended Cornell customization. I thought about adding a comment surrounding the part of the code that was added for this change but comments can also be tricky. If any other developer can suggest a better way to implement this that would be greatly appreciated!

File: src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java (hunk 6 above)
Comment (db483) [inline src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java]: Bryan, thank you very much for pointing out an example, I will update this.

File: src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java (hunk 1 above)
Comment (se12) [inline src/main/java/edu/cornell/kfs/fp/document/service/CuDisbursementVoucherTaxService.java]: Yes, that sounds better, I think makes it more readable (Clean Code!).  Thanks

**Comment by db483** [comment]
I have updated the code with the changes suggested during code review. Please let me know if you find any issues. Thank you!

File: src/main/java/edu/cornell/kfs/fp/document/validation/impl/CuDisbursementVoucherDocumentPreRules.java [hunk 14]
```diff
@@ -67,11 +69,30 @@ public boolean doPrompts(Document document) {
...
+	private void setIncomeClassNonReportableForForeignVendorWithNoTaxReviewRequired(Document document) {
+		DisbursementVoucherDocument dvDoc = (DisbursementVoucherDocument) document;
+		DisbursementVoucherPayeeDetail dvPayeeDetail = dvDoc.getDvPayeeDetail();
+		
+		String payeeTypeCode = dvPayeeDetail.getDisbursementVoucherPayeeTypeCode();
+		String paymentReasonCode = dvPayeeDetail.getDisbVchrPaymentReasonCode();
+		Integer vendorHeaderId = dvPayeeDetail.getDisbVchrVendorHeaderIdNumberAsInteger();
+
+		if (getCuDisbursementVoucherTaxService().isForeignVendorAndTaxReviewNotRequired(payeeTypeCode,paymentReasonCode, vendorHeaderId)) {
+			dvDoc.getDvNonResidentAlienTax().setIncomeClassCode(DisbursementVoucherConstants.NRA_TAX_INCOME_CLASS_NON_REPORTABLE);
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/fp/document/validation/impl/CuDisbursementVoucherDocumentPreRules.java]: I'm wondering whether the PreRules is the correct place to be setting this value, since there doesn't appear to be any routing stage or user prompt restrictions to block the update. It would be good if someone else could chime in on whether they think such updates should be moved to a different part of the code if the restrictions mentioned earlier do not apply.

File: src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+	
+	@Test
+	public void testIsForeignVendorAndTaxReviewRequired_true()  {
+		boolean isForeignVendorAndTaxReviewRequired = disbursementVoucherTaxService.isForeignVendorAndTaxReviewRequired(KFSConstants.PaymentPayeeTypes.VENDOR, PAYMENT_REASON_THAT_REQUIRES_TAX_REVIEW_FOR_FOREIGN_VENDOR, FOREIGN_VENDOR_HEADER_ID);
+        assertTrue("isForeignVendorAndTaxReviewRequired method should have returned true when input is foreign vendor and payment reason that requires tax review", isForeignVendorAndTaxReviewRequired);
+    }
+	
+	@Test
+	public void testIsForeignVendorAndTaxReviewRequired_false_PayeeTypeNotVendor()  {
+		boolean isForeignVendorAndTaxReviewRequired = disbursementVoucherTaxService.isForeignVendorAndTaxReviewRequired(KFSConstants.PaymentPayeeTypes.EMPLOYEE, PAYMENT_REASON_THAT_REQUIRES_TAX_REVIEW_FOR_FOREIGN_VENDOR, FOREIGN_VENDOR_HEADER_ID);
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java]: In the actual application, I think this method will get invoked with a null vendor header ID if the payee is not a vendor. Not sure if that needs to be reflected in this test.

File: src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+	
+	@Test
+	public void testIsForeignVendorAndTaxReviewNotRequired_true(){
+		boolean isForeignVendorAndTaxReviewNotRequired = disbursementVoucherTaxService.isForeignVendorAndTaxReviewNotRequired(KFSConstants.PaymentPayeeTypes.VENDOR, PAYMENT_REASON_THAT_DOES_NOT_REQUIRE_TAX_REVIEW_FOR_FOREIGN_VENDOR, FOREIGN_VENDOR_HEADER_ID);
+        assertTrue("isForeignVendorAndTaxReviewNotRequired method should have returned true when input is foreign vendor and payment reason that does not require tax review", isForeignVendorAndTaxReviewNotRequired);
+	}
+	
+	@Test
+	public void testIsForeignVendorAndTaxReviewNotRequired_false_PayeeTypeNotVendor(){
+		boolean isForeignVendorAndTaxReviewNotRequired = disbursementVoucherTaxService.isForeignVendorAndTaxReviewNotRequired(KFSConstants.PaymentPayeeTypes.EMPLOYEE, PAYMENT_REASON_THAT_DOES_NOT_REQUIRE_TAX_REVIEW_FOR_FOREIGN_VENDOR, FOREIGN_VENDOR_HEADER_ID);
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java]: Another test where the service is being invoked for a non-vendor payee but with a non-null vendor header ID.

**Comment by cah292** [comment]
I did find a couple minor items, but other than that the changes look good. Will wait for feedback before merging.

File: src/main/java/edu/cornell/kfs/fp/document/validation/impl/CuDisbursementVoucherDocumentPreRules.java (hunk 14 above)
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/document/validation/impl/CuDisbursementVoucherDocumentPreRules.java]: That is a good question. The income class field is set by the Tax reviewer when the document routes for tax review. In this case the doc will not route for tax review so the requirement was that a default value of 'N' is set in the income class field. The pre-rules are used for setting default values before business rules are applied so it seemed like the right place to do this. Bryan, do you have any suggestions? Thank you.

File: src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java (hunk 15 above)
Comment (db483) [inline src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java]: Good point, I will look into this and update it.

Comment (db483) [inline src/test/java/edu/cornell/kfs/fp/document/service/impl/CuDisbursementVoucherTaxServiceImplTest.java]: Good point, I will update the test

File: src/main/java/edu/cornell/kfs/fp/document/validation/impl/CuDisbursementVoucherDocumentPreRules.java (hunk 14 above)
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/fp/document/validation/impl/CuDisbursementVoucherDocumentPreRules.java]: I am ok with this code being in the PreRules. It does seem consistent with what pre-rules do in other cases such as DisbursementVoucherDocumentPreRules.checkSpecialHandlingIndicator().

**Comment by bhutchinson** [comment]
retest this please

**Comment by bhutchinson** [comment]
retest this please

**Comment by db483** [comment]
Hi Chad, I have update CuDisbursementVoucherTaxServiceImplTest.java to use null for vendorId when payee type is employee. Please review and let me know if you find any issues. Thank you.

**Comment by cah292** [comment]
The changes look good and the unit tests have passed. Merging.
